zebrad/src/application.rs

```diff
@@ -242,15 +246,22 @@ impl Application for ZebradApp {
 
     /// Load this application's configuration and initialize its components.
     fn init(&mut self, command: &Self::Cmd) -> Result<(), FrameworkError> {
+        let commit_span = error_span!("", zebrad = ZebradApp::git_commit());
+        let _commit_guard = commit_span.enter();
```

Comment:
I think the best strategy here might be to:
1. Create a ZebradApp variable for the current global span, and initialise it with the git commit hash
2. When we load the config, if it is Some, replace the global span with a span that includes the network config
3. Leak the span when the app is dropped, so it still appears in any final log messages

Should we also include the commit and network in the panic log message itself?
Maybe using wrap_err?
They're already in the URL, but we disable the URL for some panics.

Reply:
> If we're interested in adding more span info to logs emitted early during initialization we could look into installing a tracing subscriber before loading the config using some pre-configured defaults. Tracing doesn't let you overwrite the global subscriber once it's been installed but it does let you set default scoped subscribers with drop guard controlled lifetimes. We could probably use those here, though it might be messy to manage the drop guard properly. IMO it's probably not worth it but I don't really know what motivated these additional spans so I'll defer to you on making that decision.

This is a best-effort logging solution, so we know if validation and network errors are happening on mainnet or testnet.

In particular, peer errors and block heights are pretty meaningless if we don't know the network.

Reply:
> I think the best strategy here might be to:
>     1. Create a ZebradApp variable for the current global span, and initialise it with the git commit hash
>     2. When we load the config, if it is Some, replace the global span with a span that includes the network config
>     3. Leak the span when the app is dropped, so it still appears in any final log messages

I was wrong, this design results in a whole bunch of lifetime issues, because the span holds a reference to the current function.

Reply:
> Edit3: Actually, a much easier solution might be to just leak the guard. Seems to work, though we'd need to deduplicate the one in `commands.rs` if we decide to go that route

I implemented this solution, and it works except for a few spawned tasks near the start of the logs. That's good enough for now.

Reply:
> Should we also include the commit and network in the panic log message itself?
> Maybe using wrap_err?
> They're already in the URL, but we disable the URL for some panics.

I also made this change.

Reply:
> I also made this change.

wdym?

Reply:
I added a metadata section to the panic log message, in addition to the metadata in the issue URL, because we sometimes disable the issue URL. (Or people ignore it when they copy and paste the panic message.)

https://github.com/ZcashFoundation/zebra/pull/1568/files#diff-8f1ee6cceb154001084d1b17ee893fea940caa97b4c2db44f129178845b6e2f8R166